cpp/src/centrality/betweenness_centrality_impl.cuh

```diff
@@ -538,6 +542,720 @@ void accumulate_edge_results(
+    distances_2d.begin(),
+    cuda::proclaim_return_type<vertex_t>(
+      [invalid_distance] __device__(auto d) { return d == invalid_distance ? vertex_t{0} : d; }));
+  vertex_t global_max_distance = thrust::reduce(handle.get_thrust_policy(),
+                                                d_first,
+                                                d_first + distances_2d.size(),
+                                                vertex_t{0},
+                                                thrust::maximum<vertex_t>());
+
+  // PRE-COMPUTE: Partition all (vertex, source) pairs by distance ONCE
+  // This eliminates the need to scan the distance array global_max_distance times
+
+  // Count vertices at each distance level first
+  rmm::device_uvector<size_t> distance_counts(global_max_distance + 1, handle.get_stream());
+  thrust::fill(
+    handle.get_thrust_policy(), distance_counts.begin(), distance_counts.end(), size_t{0});
+
+  // Step 1: Count vertices at each distance level
+  thrust::for_each_n(
+    handle.get_thrust_policy(),
+    thrust::make_counting_iterator<size_t>(0),
+    num_sources * num_vertices,
+    [distances_2d = distances_2d.data(),
+     num_vertices,
+     distance_counts = distance_counts.data(),
+     global_max_distance] __device__(size_t global_idx) {
+      size_t source_idx         = global_idx / num_vertices;
+      vertex_t v_offset         = global_idx % num_vertices;
+      const vertex_t* distances = distances_2d + source_idx * num_vertices;
+      vertex_t dist             = distances[v_offset];
+
+      if (dist >= 0 && dist <= global_max_distance) {
+        cuda::atomic_ref<size_t, cuda::thread_scope_device> counter(distance_counts[dist]);
+        counter.fetch_add(size_t{1}, cuda::std::memory_order_relaxed);
+      }
+    });
+
+  // Copy counts to host and allocate buckets
+  std::vector<size_t> host_distance_counts(global_max_distance + 1);
+  raft::update_host(host_distance_counts.data(),
+                    distance_counts.data(),
+                    global_max_distance + 1,
+                    handle.get_stream());
+  handle.sync_stream();
+
+  // Create buckets for each distance level
+  std::vector<rmm::device_uvector<vertex_t>> distance_buckets_vertices;
+  std::vector<rmm::device_uvector<origin_t>> distance_buckets_sources;
+
+  // Reserve space for all distance levels
+  distance_buckets_vertices.reserve(global_max_distance + 1);
+  distance_buckets_sources.reserve(global_max_distance + 1);
+
+  // Allocate exact-sized buckets
+  for (vertex_t d = 0; d <= global_max_distance; ++d) {
+    if (host_distance_counts[d] > 0) {
+      distance_buckets_vertices.emplace_back(host_distance_counts[d], handle.get_stream());
+      distance_buckets_sources.emplace_back(host_distance_counts[d], handle.get_stream());
+    } else {
+      distance_buckets_vertices.emplace_back(0, handle.get_stream());
+      distance_buckets_sources.emplace_back(0, handle.get_stream());
+    }
```

Comment:
Do you need this if - else statement? The if and the else paths execute the same code.